File: epi_judge_python_solutions/defective_jugs.py
import collections
import functools
from typing import List

from test_framework import generic_test
from test_framework.test_utils import enable_executor_hook

Jug = collections.namedtuple('Jug', ('low', 'high'))


def check_feasible(jugs: List[Jug], L: int, H: int) -> bool:
    @functools.lru_cache(None)
    def check_feasible_helper(L, H):
        if L > H or (L < 0 and H < 0):
            return False

        # Checks the volume for each jug to see if it is possible.
        return any((L <= j.low and j.high <= H)
                   or check_feasible_helper(L - j.low, H - j.high)
                   for j in jugs)

    return check_feasible_helper(L, H)


@enable_executor_hook
def check_feasible_wrapper(executor, jugs, l, h):
    jugs = [Jug(*x) for x in jugs]
    return executor.run(functools.partial(check_feasible, jugs, l, h))


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('defective_jugs.py',
                                       'defective_jugs.tsv',
                                       check_feasible_wrapper))

File: epi_judge_cpp_solutions/defective_jugs.cc
#include <algorithm>
#include <memory>
#include <unordered_set>
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/serialization_traits.h"

using std::make_unique;
using std::unordered_set;
using std::vector;

struct Jug;

struct VolumeRange;
struct HashVolumeRange;

bool CheckFeasibleHelper(const vector<Jug>&, int, int,
                         unordered_set<VolumeRange, HashVolumeRange>*);

struct Jug {
  int low, high;
};

struct VolumeRange {
  int low, high;

  bool operator==(const VolumeRange& that) const {
    return low == that.low && high == that.high;
  }
};

struct HashVolumeRange {
  size_t operator()(const VolumeRange& p) const {
    return static_cast<size_t>(31) * p.low + p.high;
  }
};

bool CheckFeasible(const vector<Jug>& jugs, int L, int H) {
  return CheckFeasibleHelper(
      jugs, L, H,
      make_unique<unordered_set<VolumeRange, HashVolumeRange>>().get());
}

bool CheckFeasibleHelper(const vector<Jug>& jugs, int L, int H,
                         unordered_set<VolumeRange, HashVolumeRange>* c) {
  if (L > H || c->count({L, H}) || (L < 0 && H < 0)) {
    return false;
  }

  // Checks the volume for each jug to see if it is possible.
  if (any_of(begin(jugs), end(jugs), [&](const Jug& j) {
        return (L <= j.low && j.high <= H) ||
               CheckFeasibleHelper(jugs, L - j.low, H - j.high, c);
      })) {
    return true;
  }
  c->emplace(VolumeRange{L, H});  // Marks this as impossible.
  return false;
}

namespace test_framework {
template <>
struct SerializationTrait<Jug> : UserSerTrait<Jug, int, int> {};
}  // namespace test_framework

bool operator==(const Jug& lhs, const Jug& rhs) {
  return lhs.low == rhs.low && lhs.high == rhs.high;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"jugs", "L", "H"};
  return GenericTestMain(args, "defective_jugs.cc", "defective_jugs.tsv",
                         &CheckFeasible, DefaultComparator{}, param_names);
}
